Add Dtls Srtp support (RFC5764)

jeannotlapin opened the pull request:
## Description
This is PR supersedes PR #361 then #1540 then #1813 (discussion and review already partially performed on this one)with rebase on current dev tree

## Status
READY

## Requires Backporting
 NO  

## Migrations
NO

gilles-peskine-arm commented:
Thanks! A few things have changed recently.

Can you please add a signed-off-by line to every commit message that's yours? See [the current version of `CONTRIBUTING.md`](https://github.com/ARMmbed/mbedtls/blob/development/CONTRIBUTING.md#user-content-making-a-contribution) — in a nutshell, this indicates that you're sharing the code under the [project license](https://github.com/ARMmbed/mbedtls/blob/development/README.md#License) and that [you have the right to do so](https://github.com/ARMmbed/mbedtls/blob/development/dco.txt). This is necessary for us to accept contributions now. Please don't add signed-off-by lines to commit made by Ron unless you modify the commit (I'm not sure about the legal situation for a commit that you didn't originally write).

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -494,6 +494,10 @@ Changes
    * Ciphersuites based on 3DES now have the lowest priority by default when
      they are enabled.
 
+Features
```
Please add a file in `ChangeLog.d` instead of modifying `ChangeLog` directly. We changed this to avoid recurring add/add conflicts in `ChangeLog`.

jeannotlapin commented:
> Please don't add signed-off-by lines to commit made by Ron unless you modify the commit (I'm 
not sure about the legal situation for a commit that you didn't originally write).

Sorry I saw your comment after pushing the modified version of the PR with all commits signed off by me, including Ron's ones...

gilles-peskine-arm commented:
As far as we (the maintainers of Mbed TLS) are concerned, it's fine if you put a signed-off-by on Ron's commits. I _think_ that's ok because you're certifying that the commit is “based upon previous work that (…) is covered under an appropriate open source license”, but I cannot offer legal advice.

CI is unhappy. Mbed OS CI testing is currently broken so don't worry about that, but the other jobs need to pass. If Travis is happy but “TLS testing” isn't, ping us to let you know what's wrong (the Jenkins logs aren't publicly accessible).

mpg commented:
Hi @jeannotlapin and thanks for updating your contribution!

We (Mbed TLS maintainers) had a monthly planning meeting yesterday, and unfortunately it looks like we're unlikely to have enough bandwidth in May to fully review your PR, considering it's a pretty significant contribution (and thanks a lot for that!) and a security-sensitive feature that requires careful review.

So I'm letting you know in order to set expectations and avoid disappointments. We realise this is a contribution that started many years ago, and we're grateful that you're still willing to update it and work with us to get it integrated. We're sorry we haven't been responsive enough in the past; with the recent change in project structure we now have a much stronger focus on working with the community, but we also find ourselves with a pretty large backlog to handle so it's going to take some time.

We really hope to be able to allocate the necessary review time in the near future and finally merge this long-standing contribution in a reasonable delay, but unfortunately given our current workload it probably won't happen quite this month.

Thanks for you understanding and patience.

lherman-cs commented:
This is awesome! I'm working on an opensource WebRTC library written in C, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c, and currently enabling the ability to use mbedTLS. And, this is the only missing piece that we need to completely swap out OpenSSL with mbedTLS, it's required to enable DTLS-SRTP extension.

Do you guys have any update on this?

ramatin commented:
So close... Have been waiting eagerly to replace OpenSSL for WebRTC use case. You guys have put in a lot of efforts. Thanks...

ycyang1229 commented:
@mpg @gilles-peskine-arm Hey, recently, I am working on the webrtc development on arm embedded devices, and I think this feature is quite critical. I saw the latest response from the mbedtls team was in Apr. I wonder is there any chance mbedtls can provide similar features or merge this pull req? I truly hope mbedtls can have similar feature, and it will help the webrtc lands on ipc industry. 
thanks.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     * Min length is 5: at least one protection profile(2 bytes)
+     *                  and length(2 bytes) + srtp_mki length(1 byte)
+     */
+    if( len < size_of_lengths + 2 )
```
This non-local bounds check is hard to read, and considering 2 Bytes for at least one protection profile is redundant since there's another check for the actual protection profile list length below.

Suggestion: Do one check a time, i.e. (a) check that `len >= 2` here, (b) read protection profile list length, (c) check that extension length is big enough to hold protection profile list, (d) parse profile list, etc, ... 

Checking for space for the MKI length byte could be added to (c), though.

jeannotlapin replied:
done

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+        uint16_t protection_profile_value = buf[j + 2] << 8 | buf[j+3];
+        client_protection = mbedtls_ssl_get_srtp_profile_value( protection_profile_value );
+
+        profile_info = mbedtls_ssl_dtls_srtp_profile_info_from_id( client_protection );
```
Suggestion: For efficiency, `continue` when `profile_info == NULL`.

jeannotlapin replied:
done

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+    if( ( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED ) &&
+        ( len > ( profile_length + 2 ) ) )
+    {
+        ssl->dtls_srtp_info.mki_len = buf[profile_length + 2];
```
Blocker: To my understanding, the MKI field -- consisting of length + value -- is _always_ present in the extension, but its value may have length `0`, indicating that an MKI should not be used. Therefore, we must unconditionally parse the field here, regardless of the setting of `MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED`.

jeannotlapin replied:
Correct, I moved the parsing of mki_length and checking of global len out of the MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED if and keep it only for the parsing of the mki itself.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+        }
+
+        MBEDTLS_SSL_DEBUG_BUF( 3, "using mki",  ssl->dtls_srtp_info.mki_value,
+                                                ssl->dtls_srtp_info.mki_len );
```
Blocker: We must match the total extension length against the cumulative length of the fields found in the extension.

jeannotlapin replied:
done

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+    for( j=0; j < profile_length; j += 2 )
+    {
+        /* + 2 to skip the length field */
+        uint16_t protection_profile_value = buf[j + 2] << 8 | buf[j+3];
```
Suggestion: Improve readability by gradually incrementing `buf`.

jeannotlapin replied:
done

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -2499,6 +2612,61 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    }
+
+    buf[8] = mki_len & 0xFF;
+    memcpy( &buf[9], ssl->dtls_srtp_info.mki_value, mki_len );
```
Blocker: It's a deficiency of our current extension-writing interface that we don't pass an end-of-buffer field to bounds-check the extension writing functions against. That's normally not an issue for statically sized extensions, assuming that Mbed TLS isn't configured with extremely small buffer sizes, but in the case at hand, we're echoing the client-provided MKI value which has variable length and thereby has the potential to silently overflow the write buffer. 

Suggestion: Add an `end` pointer to the extension writing function and add an explicit bounds check.

jeannotlapin replied:
This one is a bit harder than the others. All the extension writing functions shall be rewritten.
I can't see in the ssl_write_server_hello function a way to access the writing buffer size. Is there a constant that define this buffer size somewhere?
The client hello extension is variable but not unbounded, its max size is 255 bytes.

jeannotlapin replied:
@hanno-arm any hint to get the actual max buffer size?

mpg replied:
Unless I'm mistaken, this days one should use `ssl->out_msg + MBEDTLS_SSL_OUT_CONTENT_LEN` to get the end of the buffer (or more precisely, the limit usable for record contents - the actual buffer is a bit larger in order to allow for record protection overhead).

jeannotlapin replied:
Thanks @mpg I added the check on buffer size as done on the `ssl_write_cid_ext` using  `ssl->out_msg + MBEDTLS_SSL_OUT_CONTENT_LEN` as you suggested.

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -409,6 +414,15 @@
+ * Use_srtp extension protection profiles values as defined in
+ * http://www.iana.org/assignments/srtp-protection/srtp-protection.xhtml
+ */
+#define MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80_IANA_VALUE     0x0001
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +865,43 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+
+#define MBEDTLS_DTLS_SRTP_MAX_KEY_MATERIAL_LENGTH    60
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *                          in decreasing preference order.
+ * \param profiles_number   Number of supported profiles.
+ *
+ * \return                  0 on success, or #MBEDTLS_ERR_SSL_BAD_INPUT_DATA.
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ */
+int mbedtls_ssl_conf_dtls_srtp_protection_profiles
+                               ( mbedtls_ssl_config *conf,
+                                 const mbedtls_ssl_srtp_profile *profiles,
```
The parameter is an array of mbedtls_ssl_srtp_profile which is an enum, so a compiler would not accept a 0 value.
But MBEDTLS_SRTP_UNSET_PROFILE shall be mapped to 0. I can switch to a MBEDTLS_SRTP_UNSET_PROFILE terminated array for the profiles parameter and remove the profiles_number parameter.

gilles-peskine-arm replied:
Generally I prefer pointer+length to a zero-terminated array because it's easier to keep track of the buffer size. With a length: allocate N elements, pass N together with the pointer everywhere. With zero termination, you have to remember to create the trailing zero, and if the array isn't constant you also have to ensure that the creation doesn't overflow the array.

However consistency with existing zero-terminated lists in the interface is a valid argument. So I don't have strong feelings on this one.

jeannotlapin replied:
@hanno-arm  any comment? Shall I let it the way it is or change for the MBEDTLS_SRTP_UNSET_PROFILE-terminated list?

hanno-arm replied:
I'd prefer an `MBEDTLS_SRTP_UNSET_PROFILE`-terminated list to be aligned with the rest of the API.

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+/**
+ * \brief                   Add support for mki value in use_srtp extension.
```
done

hanno-arm replied:
Thanks!

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ * \param mki_value        The MKI value to set.
+ * \param mki_len          The length of the MKI value.
+ *
+ * \return         0 on success, #MBEDTLS_ERR_SSL_BAD_INPUT_DATA
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *
+ * \param ssl      SSL context
+ *
+ * \return         Protection Profile enum member,
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *
+ * \param profile          The dtls-srtp profile id to get info on.
+ *
+ * \return                 Address of the SRTP profile information structure on
```
done

hanno-arm replied:
Style: Full stop missing, replace "Address of" by "The address of".
Style: Overly long line.

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+/**
+ * \brief                  Utility function to get information on DTLS-SRTP profile.
+ *
+ * \param profile          The dtls-srtp profile id to get info on.
```
done

jeannotlapin commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+
+    /* If use_srtp is not configured, just ignore the extension */
+    if( ssl->conf->dtls_srtp_profile_list == NULL ||
+        ssl->conf->dtls_srtp_profile_list_len == 0 )
```
done

jeannotlapin commented on library/ssl_cli.c:
```diff
@@ -3246,8 +3469,20 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
     {
-        /* Current message is probably the ServerHelloDone */
-        ssl->keep_current_message = 1;
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
```
You're right, it seems that Ron's understanding of the RFC was different. I removed all this check on the client side.

jeannotlapin commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,110 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+   ssl->dtls_srtp_info.chosen_dtls_srtp_profile = MBEDTLS_SRTP_UNSET_PROFILE;
+
+    /* first 2 bytes are protection profile length(in bytes) */
+    profile_length = ( buf[0] << 8 ) | buf[1];
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ * \param profile          The dtls-srtp profile id to get info on.
+ *
+ * \return                 Address of the SRTP profile information structure on
+ *                         success,NULL if not found.
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ * \param ssl      SSL context
+ *
+ * \return         Protection Profile enum member,
+ *                 #MBEDTLS_SRTP_UNSET_PROFILE if no protocol was negotiated.
```
done

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *                 This function should be called after the handshake is
+ *                 completed.
+ *
+ * \param ssl      SSL context
```
done

hanno-arm replied:
There are still cases of this remaining it seems.

jeannotlapin commented:
@hanno-arm I fixed all the blockers, the minor API consistency is waiting for you to confirm the modification or not.
The travis CI build is failing to something which does not seem related to this PR:
https://travis-ci.org/github/ARMmbed/mbedtls/jobs/723371507

`ssl_msg.c: In function 'mbedtls_ssl_cf_memcpy_if_eq':

ssl_msg.c:1085:5: error: 'for' loop initial declarations are only allowed in C99 or C11 mode

     for( size_t i = 0; i < len; i++ )

     ^

ssl_msg.c:1085:5: note: use option -std=c99, -std=gnu99, -std=c11 or -std=gnu11 to compile your code

Makefile:263: recipe for target 'ssl_msg.o' failed

make[1]: *** [ssl_msg.o] Error 1

make[1]: *** Waiting for unfinished jobs....

Makefile:17: recipe for target 'lib' failed

make: *** [lib] Error 2`

gilles-peskine-arm commented:
We just fixed the Travis failure (https://github.com/ARMmbed/mbedtls/pull/3632). It's going to take a few hours for Travis to go through all the pull requests again.

hanno-arm commented:
@jeannotlapin Thanks for your rework, I'm back from holiday and will make time this week to re-review this PR.

jeannotlapin commented:
Hi @hanno-arm 
any chance to review the modification you requested ?

hanno-arm commented:
@jeannotlapin Yes, I aim to do another round of re-review either later today or tomorrow.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -3210,7 +3212,8 @@ void mbedtls_ssl_conf_srtp_mki_value_supported( mbedtls_ssl_config *conf,
  *
- * \return                  0 on success, or #MBEDTLS_ERR_SSL_BAD_INPUT_DATA.
+ * \return                  0 on success
+ * \return                  #MBEDTLS_ERR_SSL_BAD_INPUT_DATA when the list of protection profiles is incorrect
```
Minor: Overly long line. 
Minor: Full stop missing at the end.

jeannotlapin replied:
fixed

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,79 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+/**
+ * \brief                   Add support for mki(master key id) value in use_srtp extension.
+ *                          MKI is an optional part of SRTP used for key management and
+ *                          re-keying. See RFC3711 section 3.1 for details
```
Style: Full stop missing.

jeannotlapin replied:
fixed

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,79 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *                          The default value is
+ *                          #MBEDTLS_SSL_DTLS_SRTP_MKI_UNSUPPORTED.
+ *
+ * \param conf              SSL configuration
```
Minor, as earlier: Try to use full sentences in the Doxygen documentation.

jeannotlapin replied:
fixed

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,79 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *
+ * \return                 Address of the SRTP profile information structure on
+ *                         success
+ * \return                 \c NULL if not found.
```
Minor: `\c NULL if the protection profile \p profile was not found.`

jeannotlapin replied:
fixed

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +865,43 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
+
+typedef struct mbedtls_dtls_srtp_info_t
+{
+    /*! The SRTP profile that was negotiated*/
```
Minor: Space missing before `*/`.

jeannotlapin replied:
fixed

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +865,43 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
+    /*! The mki_value used, with max size of 256 bytes */
+    unsigned char mki_value[MBEDTLS_TLS_SRTP_MAX_MKI_LENGTH];
+    /*! The length of mki_value */
+    size_t                 mki_len;
```
Style: Odd spacing.

jeannotlapin replied:
fixed

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4685,6 +4714,106 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    return( 0 );
+}
+
+int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
```
In line with existing API, e.g. the configuration of SSL ciphersuites or the configuration of the ALPN extension, I'd suggest not to copy the list of protection profiles here, but remember the user-provided pointer. This requires the caller to ensure that the profile list remains valid for the lifetime of the SSL configuration, which needs to be appropriately documented - see e.g. the documentations of `mbedtls_ssl_conf_ciphersuite()` or `mbedtls_ssl_conf_alpn_protocols()`.

jeannotlapin replied:
done

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +865,43 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
+/*
+ * List of SRTP profiles for DTLS-SRTP
+ */
+typedef enum
```
On second look, I'm not sure this enum is necessary, as we might as well use the named IANA constants throughout the code, and `mbedtls_ssl_srtp_profile` could be replaced by `typdef uint16_t mbedtls_ssl_srtp_profile` to still have a type name to indicate the semantics of those numeric values.

hanno-arm replied:
Thoughts, @gilles-peskine-arm @mpg @jeannotlapin ?

gilles-peskine-arm replied:
Why not define the enum with the same values as the IANA constants?

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4685,6 +4714,106 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
 #endif /* MBEDTLS_SSL_ALPN */
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+static const mbedtls_ssl_srtp_profile_info srtp_profile_definitions[] =
```
Design question, ping @mpg @gilles-peskine-arm: Do we want to export those strings unconditionally, even if they're for debugging purpose only? 

I prefer not do because it looks like a waste of ROM in constrained environments, but on the other hand, existing information structures, e.g. for Cipher or MD, do also contain a string representation of the respective object (which in the latter cases are much smaller, though).

gilles-peskine-arm replied:
We've had complaints about too many strings values, although I'm not sure if it was about debugging code: the one I remember is that if you include anything that links to `oid.c`, you get all the OID. I think we should include them only if `MBEDTLS_DEBUG_C` is defined.

gilles-peskine-arm replied:
Sounds good to me.

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -253,10 +256,26 @@ int main( void )
+    "    use_srtp=%%d         default: 0 (disabled)\n" \
+    "                          This cannot be used with eap_tls=1 or "\
+    "                          nss_keylog=1\n"             \
+    "    srtp_force_profile=%%d  default: all enabled\n"   \
```
Which numeric value selects this default behaviour - `0`? If so, it might be worth adding that: `default: all enabled (0)`.

gilles-peskine-arm replied:
It should be `default: 0 (all enabled)` for consistency. That's not very important though, the documentation of many options is incomplete anyway.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -653,7 +676,43 @@ static int nss_keylog_export( void *p_expkey,
+
+    if( opt.debug_level > 2 )
+    {
+        mbedtls_printf("exported maclen is %u\n", (unsigned)maclen);
```
Very minor: Space after `(` and before `)`.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -1174,7 +1237,13 @@ int main( int argc, char *argv[] )
     eap_tls_keys eap_tls_keying;
-#endif
+#if defined( MBEDTLS_SSL_DTLS_SRTP )
+    /*! master keys and master salt for SRTP generated during handshake */
```
Style: Indentation.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2234,6 +2317,44 @@ int main( int argc, char *argv[] )
+        }
+        else
+        {
+            const mbedtls_ssl_srtp_profile default_profiles[] =
```
Note: This needs to be at function scope if we change the `mbedtls_ssl_conf_dtls_srtp_protection_profiles()` API to not make a copy of the profile list.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2234,6 +2317,44 @@ int main( int argc, char *argv[] )
+
+        if( ret != 0 )
+        {
+            mbedtls_printf( " failed\n  ! mbedtls_ssl_conf_dtls_srtp_protection_profiles returned %d\n\n", ret );
```
Minor: Overly long line.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2234,6 +2317,44 @@ int main( int argc, char *argv[] )
 #endif
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    if( opt.use_srtp != DFL_USE_SRTP )
```
Minor: This forces the default value to be `0`/off - it wouldn't work anymore if it changed to `1`/on.

Suggestion: Modify the code to be agnostic of the default values for command line parameters.

hanno-arm replied:
This is a defect with pre-existing command line parameters, but that shouldn't prevent us from doing it right when introducing new ones.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2470,6 +2598,26 @@ int main( int argc, char *argv[] )
 #endif
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    if( opt.use_srtp != DFL_USE_SRTP &&  strlen( opt.mki ) != 0 )
```
Minor: Double space after `&&`.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2470,6 +2598,26 @@ int main( int argc, char *argv[] )
+                                   opt.mki,&mki_len ) != 0 )
+        {
+            mbedtls_printf( "mki value not valid hex\n" );
+             goto exit;
```
Minor: Spacing.

jeannotlapin replied:
fixed

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2591,7 +2739,39 @@ int main( int argc, char *argv[] )
+    {
+        size_t j = 0;
+
+        if( ( ret = mbedtls_ssl_tls_prf( dtls_srtp_keying.tls_prf_type,
```
I don't think we've checked at this point that DTLS SRTP has actually been negotiated, or did we?

jeannotlapin replied:
I added a check : fetch the negotiated protection profile. As newly explained in the doc, it is available as soon as the server Hello as been parsed.

hanno-arm commented on programs/ssl/ssl_server2.c:
```diff
@@ -324,10 +327,24 @@ int main( void )
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+#define USAGE_SRTP \
+    "    use_srtp=%%d         default: 0 (disabled)\n" \
+    "    srtp_force_profile=%%d  default: all enabled\n"   \
```
Same as above: `all enabled (0)`.

jeannotlapin replied:
done

hanno-arm commented on programs/ssl/ssl_server2.c:
```diff
@@ -758,7 +779,43 @@ static int nss_keylog_export( void *p_expkey,
+
+    if( opt.debug_level > 2 )
+    {
+        mbedtls_printf("exported maclen is %u\n", (unsigned)maclen);
```
Same comments as for `ssl_client2.c`. I'm not going repeating all of them.

jeannotlapin replied:
done

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -756,6 +756,107 @@ static int ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+     * SRTPProtectionProfile SRTPProtectionProfiles<2..2^16-1>;
+     */
+    if( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED &&
+        ssl->dtls_srtp_info.mki_len != 0 )
```
What's the relevant of the check `ssl->dtls_srtp_info.mki_len != 0` here?

jeannotlapin replied:
you're right, it was useless. I removed it

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -2500,6 +2627,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+        mki_len = ssl->dtls_srtp_info.mki_len;
+    }
+
+    if( end < buf + mki_len + 9 )
```
Minor: Explain the magic `9` in a comment.

jeannotlapin replied:
done

hanno-arm reviewed (commented):
@jeannotlapin I haven't finished the ongoing review, but here's a first bunch of comments.

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,79 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ * \param ssl      The SSL context to query
+ *
+ * \return         The DTLS SRTP protection profile in use
+ * \return         #MBEDTLS_SRTP_UNSET_PROFILE if no protocol was negotiated or the handshake is still on
```
fixed

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3188,79 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *
+ * \return         The DTLS SRTP protection profile in use
+ * \return         #MBEDTLS_SRTP_UNSET_PROFILE if no protocol was negotiated or the handshake is still on
+ *                 early stage
```
The negotiated profile is actually available as soon as the peer's Hello packet has been parsed, I updated the doc to specify that.

jeannotlapin commented:
@hanno-arm I addressed most of your comments.

The mess with the enum, its IANA value and associated string is pending.
What about : 
- keeping the enum, setting its values to the IANA one(having MBEDTLS_SRTP_UNSET_PROFILE to 0)
- having a function returning the string value defined only when MBEDTLS_DEBUG_C is defined
- having the MBEDTLS_SRTP_UNSET_PROFILE terminated list as input of mbedtls_ssl_conf_dtls_srtp_protection_profiles?

Would that close it all?

hanno-arm reviewed (changes requested):
Except for trivialities, the following needs addressing:
* I suggest to remove the DTLS SRTP protection profile enum and exclusively work with the IANA values. @gilles-peskine-arm suggested the alternative of using an enum with the IANA constants, but I think that would deviate from existing code defining IANA constants as macros. This would also remove the need for conversion functions. You could still introduce a `typedef` for the protection profiles for readability.
* I suggest to remove the DTLS SRTP info structure and instead use a helper, guarded by `MBEDTLS_DEBUG_C`, which provides a string presentation of DTLS SRTP protection profiles.
* I suggest to make the protection profile list terminated by `MBEDTLS_SRTP_UNSET_PROFILE` and remove the length parameter, in line with the API for e.g. ciphersuites or the ALPN extension.
* A bounds check in the use_srtp writing function needs to be rewritten.

In order to avoid unnecessary back and forth, I'd like another opinion on the first three points, though. Ping @gilles-peskine-arm @mpg @yanesca

hanno-arm commented:
@jeannotlapin A general note on updating this PR: It would help the review process if we'd avoid forced pushes and leave a potential history cleanup to the end of the review process. Also, I personally prefer to resolve the conversations that I opened during the review process, as it helps me tracking which comments I have checked to be addressed, and which are still pending.

This is only a suggestion and by no means criticism nor a justification for the ridiculous time this PR is already open. Thank you again for your work and patience.

hanno-arm commented:
@jeannotlapin I forgot to mention one important point in the review summary: As it stands, the PR still enforces the use of server authentication when DTLS SRTP is in use. What's the rationale for that? I'd prefer to not entangle SRTP with the authentication configuration unless the RFC clearly says so -- and if it does, I'd suggest to fail cleanly if a user attempts to negotiate DTLS-SRTP while disabling server authentication, instead of silently overwriting the authentication configuration.

jeannotlapin commented:
> @jeannotlapin A general note on updating this PR: It would help the review process if we'd avoid forced pushes and leave a potential history cleanup to the end of the review process. Also, I personally prefer to resolve the conversations that I opened during the review process, as it helps me tracking which comments I have checked to be addressed, and which are still pending.

@hanno-arm Ok, I won't close the discussion, it was also for me a way to track what I've closed. Most of forced pushes are to fix minor problems detected by the build checks (forgetting to sign-off, having some tabs in the code....) so it would just make the commit history messier to keep them. Some of them are for rebase, same purpose, it is much cleaner than ugly merges.

hanno-arm commented:
@jeannotlapin Agreed, we don't want merge commits in the PR history, but rebases can harden the review. My preference is to accept a lengthy history during the review process by pushing fixup commits, and potentially clean it up with a zero-diff force push prior to merging.

jeannotlapin commented:
> @jeannotlapin I forgot to mention one important point in the review summary: As it stands, the PR still enforces the use of server authentication when DTLS SRTP is in use. What's the rationale for that? I'd prefer to not entangle SRTP with the authentication configuration unless the RFC clearly says so -- and if it does, I'd suggest to fail cleanly if a user attempts to negotiate DTLS-SRTP while disabling server authentication, instead of silently overwriting the authentication configuration.

@hanno-arm The RFC clearly states that in presence of the use_srtp extension in the client Hello the server will add its certificate and a certificate request to its response to it : See [section 4.1](https://tools.ietf.org/html/rfc5764#page-6). I remember you spotted the fact that the RFC uses will and not MUST so it is unclear.

In the context of DTLS-SRTP exchange, there is not much sense in performing the DTLS handshake without exchanging both certificates (but not enforcing their verification at handshake level), this bypassing lazy configuration is to help the library users to make a correct use of the functionality but I can remove it and/or keep the check and abort the handshake. Let me know.

mpg commented:
> Ok, I won't close the discussion, it was also for me a way to track what I've closed.

Note: we tend to use reaction emojis on comments for that. There's no need for a shared convention because it's just for your own tracking, but what I (and others in the team) tend to do is :+1: for "ack, will fix" and :rocket: for "done" - so when I push a new version, I can check that each comment either has a rocket or a reply indicating why I didn't do that change (yet).

hanno-arm commented:
> @hanno-arm The RFC clearly states that in presence of the use_srtp extension in the client Hello the server will add its certificate and a certificate request to its response to it : See section 4.1. I remember you spotted the fact that the RFC uses will and not MUST so it is unclear.

Given the ambiguity of the RFC, I prefer to proceed as for client authentication, that is, to use the authentication mechanisms that the user has configured and to not interfere with it.

mpg commented:
> I suggest to remove the DTLS SRTP protection profile enum and exclusively work with the IANA values. @gilles-peskine-arm suggested the alternative of using an enum with the IANA constants, but I think that would deviate from existing code defining IANA constants as macros. This would also remove the need for conversion functions. You could still introduce a `typedef` for the protection profiles for readability.

I fully agree that we should avoid having two way of referring to protection profiles with constants, because it creates the need to convert between those, which makes the code larger. So we should either have an enum directly with the IANA values, or a set of macros with the values, preferably with a typedef. I don't have a strong opinion between those two, but a slight inclination towards the macros+typedef, which give use more control over which type is used (an enum would make the compile choose for us in ways we can't really control).

> * I suggest to remove the DTLS SRTP info structure and instead use a helper, guarded by `MBEDTLS_DEBUG_C`, which provides a string presentation of DTLS SRTP protection profiles.

I fully agree here, if the strings are only used for debugging, we shouldn't be paying the price of having them in the binary for non-debug builds.

> * I suggest to make the protection profile list terminated by `MBEDTLS_SRTP_UNSET_PROFILE` and remove the length parameter, in line with the API for e.g. ciphersuites or the ALPN extension.

I don't have a strong opinion here.

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1815,6 +1815,20 @@
+ *
+ * Enable support for DTLS-SRTP, RFC5764, use_srtp extension.
+ * \note Only the dtls-srtp key material negotiation is supported.
+ * Once negotiated, the key should be extracted, and data should be transmitted
```
Suggestion: A comment on how the key material can be extracted would be helpful.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +869,34 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
+    /*! The mki_value used, with max size of 256 bytes. */
+    unsigned char mki_value[MBEDTLS_TLS_SRTP_MAX_MKI_LENGTH];
+    /*! The length of mki_value. */
+    size_t mki_len;
```
Considering the maximum MKI length, `mki_len` could be a `uint16_t` placed afer the `chosen_dtls_srtp_profile` to avoid padding bytes.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4685,6 +4689,93 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+
+    /* check the profiles list: all entry must be valid,
+     * its size cannot be more than the total number of supported profiles, currently 4 */
+    for( p = profiles; *p != MBEDTLS_TLS_SRTP_UNSET && list_size < 5; p++ )
```
What is the reason to restrict the list length to a maximum of `4` here? I see that there's no value in having longer lists, but the code would handle them correctly, and dropping the length check would save a few bytes of code.

jeannotlapin replied:
If you provide a list longer than 4 elements, it means you repeated some element, the order in thus not really defined, which is not really a problem, user shall use correctly the library but the interface should help him too.

Most important is to prevent an infinite loop, or wild memory browsing if the user did not carefully read the instructions and forgot the MBEDTLS_TLS_SRTP_UNSET at the end of its list, some limit must be set anyway, 4 seems to be a good one as you have no real reason to use a longer list.

hanno-arm replied:
On the other hand, the knowledge that the list length is small would allow to hardcode the high byte of the profile list length in https://github.com/ARMmbed/mbedtls/pull/3235/files#diff-8d3acce410d251c6ead39efd401db23eR805 to `0`.

In any case, this turns into micro-optimization. My preference would be to allow a maximum list length of 255 here and hardcode the high byte in https://github.com/ARMmbed/mbedtls/pull/3235/files#diff-8d3acce410d251c6ead39efd401db23eR805 to `0`, but I leave it to you to decide whether you want to change the code or not. It's not a blocker.

hanno-arm replied:
If you decide to keep the code as-is, I'd prefer giving the magic values `4` and `5` here a descriptive name, like `MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH` or similar.

hanno-arm replied:
> If you provide a list longer than 4 elements, it means you repeated some element, the order in thus not really defined, which is not really a problem, user shall use correctly the library but the interface should help him too.
> Most important is to prevent an infinite loop, or wild memory browsing if the user did not carefully read the instructions and forgot the MBEDTLS_TLS_SRTP_UNSET at the end of its list, some limit must be set anyway, 4 seems to be a good one as you have no real reason to use a longer list.

I agree that this is a reasonable way to prevent the user from shooting itself in the foot, but one has to balance such measures with their code size cost. In this instance, the cost is negligible, so I'm ok leaving it. Note, though, that there's a precedent for this, namely the ciphersuite and ALPN configuration API, and that API doesn't do any sanity checking on the list length or content - it's the user's responsibility to get it right.

jeannotlapin replied:
Ok, I kept the check and defined the MAX_PROFILE_LIST_LENGTH.

I also added the micro optimization of writing 0 to the upper byte of the list in the client use_srtp_ext writing code. Note that if you allow the max_length to be more than 127(I kept it at 4), you will need to correctly write the upper byte of the length as you're actually write 2*length in the use_srtp extension

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -1277,6 +1377,14 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+    {
+        ssl_write_use_srtp_ext( ssl, p + 2 + ext_len, &olen );
```
In line with the pattern for other extension writing functions, could you pass the `end` pointer as the last parameter and include appropriate bounds checks in `ssl_write_use_srtp_ext`?

hanno-arm replied:
NB: The code doesn't seem to be uniform in that regard: Client-side extension writing functions accept a end-of-buffer parameter, while server-side extension writing functions don't.

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -756,6 +756,106 @@ static int ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+             * Note: we shall never arrive here as protection profiles
+             * is checked by ssl_set_dtls_srtp_protection_profiles function
+             */
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, ignore illegal DTLS-SRTP protection profile %d",
```
Style: Overly long line

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -756,6 +756,106 @@ static int ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    *p++ = (unsigned char)( ext_len & 0xFF );
+
+    /* protection profile length: 2*(ssl->conf->dtls_srtp_profile_list_len) */
+    *p++ = (unsigned char)( ( ( 2 * (ssl->conf->dtls_srtp_profile_list_len) )
```
Minor: Remove parentheses around `dtls_srtp_profile_list_len`.

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -756,6 +756,106 @@ static int ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    /* protection profile length: 2*(ssl->conf->dtls_srtp_profile_list_len) */
+    *p++ = (unsigned char)( ( ( 2 * (ssl->conf->dtls_srtp_profile_list_len) )
+                              >> 8 ) & 0xFF );
+    *p++ = (unsigned char)( ( 2 * (ssl->conf->dtls_srtp_profile_list_len) )
```
Minor: Remove parentheses around `dtls_srtp_profile_list_len`.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4685,6 +4689,93 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+#if defined(MBEDTLS_DEBUG_C)
+const char *mbedtls_ssl_get_srtp_profile_as_string ( mbedtls_ssl_srtp_profile profile )
```
Blocker: This requires a declaration in `ssl.h`. Alternatively, you could make it `static inline` function, declared and defined in `ssl.h`.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4685,6 +4689,93 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    switch( profile )
+    {
+        case MBEDTLS_TLS_SRTP_AES128_CM_HMAC_SHA1_80:
+            return "MBEDTLS_TLS_SRTP_AES128_CM_HMAC_SHA1_80";
```
Style, here and in the next lines: Parentheses missing.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -3229,6 +3230,8 @@ void mbedtls_ssl_conf_srtp_mki_value_supported( mbedtls_ssl_config *conf,
  *                          for later reference as required, so the lifetime
  *                          of the table must be at least as long as the lifetime
  *                          of the SSL configuration structure.
+ *                          The list must not hold more than
+ *                          MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH elements
```
```suggestion
 *                          The list must not hold more than
 *                          MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH elements
 *                          (excluding the terminating MBEDTLS_TLS_SRTP_UNSET).
```

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4734,11 +4734,11 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
     }
 
-    if ( list_size > 4 ) {
+    if ( list_size > MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH ) {
```
```suggestion
    if ( list_size > MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH ) 
    {
```

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4734,11 +4734,11 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
                 break;
             default: /* unsupported value, stop parsing and set the size to an error value */
-                list_size = 5;
+                list_size = MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH+1;
```
```suggestion
                list_size = MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH + 1;
```

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4734,11 +4734,11 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
                     list_size++;
```
```suggestion
                list_size++;
```

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4723,7 +4723,7 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
     /* check the profiles list: all entry must be valid,
      * its size cannot be more than the total number of supported profiles, currently 4 */
-    for( p = profiles; *p != MBEDTLS_TLS_SRTP_UNSET && list_size < 5; p++ )
+    for( p = profiles; *p != MBEDTLS_TLS_SRTP_UNSET && list_size <= MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH; p++ )
```
```suggestion
    for( p = profiles; *p != MBEDTLS_TLS_SRTP_UNSET && 
                       list_size <= MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH; 
         p++ )
```

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -757,8 +757,10 @@ static int ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
-static void ssl_write_use_srtp_ext( mbedtls_ssl_context *ssl,
-                                    unsigned char *buf, size_t *olen )
+static int ssl_write_use_srtp_ext( mbedtls_ssl_context *ssl,
+                                    unsigned char *buf,
+                                    const unsigned char *end,
+                                    size_t *olen )
```
```suggestion
                                   unsigned char *buf,
                                   const unsigned char *end,
                                   size_t *olen )
```

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -1387,7 +1398,12 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
     if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
     {
-        ssl_write_use_srtp_ext( ssl, p + 2 + ext_len, &olen );
+        if( ( ret = ssl_write_use_srtp_ext( ssl, p + 2 + ext_len,
+                                       end, &olen ) ) != 0 )
```
```suggestion
        if( ( ret = ssl_write_use_srtp_ext( ssl, p + 2 + ext_len,
                                            end, &olen ) ) != 0 )
```

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1815,6 +1815,20 @@
+/**
+ * \def MBEDTLS_SSL_DTLS_SRTP
+ *
+ * Enable support for DTLS-SRTP, RFC5764, use_srtp extension.
+ * \note Only the dtls-srtp key material negotiation is supported.
+ * Once negotiated, the key should be extracted using mbedtls_ssl_tls_prf()
+ * and data should be transmitted via an SRTP stack.
```
```suggestion
 * Enable support for negotation of DTLS SRTP (RFC 5764) through the use_srtp extension.
 * 
 * This feature only implements the minimum functionality required to negotiate 
 * the use of DTLS SRTP and to allow the derivation of the associated SRTP 
 * packet protection key material.
 * In particular, the SRTP packet protection itself, as well as the demultiplexing
 * of RTP and DTLS packets at the datagram layer (see Section 5 of RFC 5764),
 * are not handled by this feature. Instead, after successful completion
 * of a handshake negotiating the use of DTLS SRTP, the extended key exporter API 
 * mbedtls_ssl_conf_export_keys_ext_cb() should be used to implement the key exporter
 * described in Section 4.2 of RFC 5764 and RFC 5705. The resulting key should then
 * be passed to an SRTP stack. 
```

hanno-arm reviewed (commented):
Apart from trivial style fixes, I left a suggestion to expand the documentation on what the new feature provides, and - importantly - what not. Let me know if you're OK with it @jeannotlapin or if you have suggestions for further improvement.

Beyond that, the PR looks good, and I'm happy with the test coverage in `ssl-opt.sh` in principle. However, what seems to be missing from those tests is the check that the SRTP keying material derivation on client and server lead to the same results. Did you investigate this? It won't be possible with the current `run_test` command which only lets you grep client and server output independently, but on first sight it shouldn't be difficult to extend the command with the ability to extract-and-compare two strings from client and server logs -- in this case, the DTLS SRTP keying material.

@gilles-peskine-arm @mpg Have you considered this before?

jeannotlapin commented:
> Apart from trivial style fixes, I left a suggestion to expand the documentation on what the new feature provides, and - importantly - what not. Let me know if you're OK with it @jeannotlapin or if you have suggestions for further improvement.

That's perfect for me. Your expanded documentation is much better than what was there before. Can you directly push your changes or shall I do it?

> 
> Beyond that, the PR looks good, and I'm happy with the test coverage in `ssl-opt.sh` in principle. However, what seems to be missing from those tests is the check that the SRTP keying material derivation on client and server lead to the same results. Did you investigate this? It won't be possible with the current `run_test` command which only lets you grep client and server output independently, but on first sight it shouldn't be difficult to extend the command with the ability to extract-and-compare two strings from client and server logs -- in this case, the DTLS SRTP keying material.

Yes this is missing in the interop test and would be great. Previous version of this PR (before the use of the export key callback) has been tested and is able to interoperate with openssl. I never tested the current version. I'll check if the openssl s_server or s_client command is able to produce logs including the DTLS-SRTP key material.

hanno-arm commented:
@jeannotlapin  
> That's perfect for me. Your expanded documentation is much better than what was there before. Can you directly push your changes or shall I do it?
Yes, please go ahead :+1:

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2470,6 +2596,26 @@ int main( int argc, char *argv[] )
+
+        mbedtls_ssl_conf_srtp_mki_value_supported( &conf, MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED );
+        if( ( ret = mbedtls_ssl_dtls_srtp_set_mki_value( &ssl, mki,
+                                                         strlen( opt.mki ) / 2 ) ) != 0 )
```
MSVC is complaining because this includes an implicit conversion from `size_t` to `uint16_t`. Could you add an explicit cast?

jeannotlapin commented:
> Yes this is missing in the interop test and would be great. Previous version of this PR (before the use of the export key callback) has been tested and is able to interoperate with openssl. I never tested the current version. I'll check if the openssl s_server or s_client command is able to produce logs including the DTLS-SRTP key material.

@hanno-arm I checked, and the server client of openssl is able to generate a trace of the exported key material. The format is something like:
`Keying material: 75B35B0D33528FF66D6135AB70EB67E415FA555C1010E55A726182384980DD98DC3ACFE4A2B0ACF8DF5C666F72245FF31C1497D029964075F9F29DAB`

I can change the format, or add trace, to get a similar line in the log of ssl_server2.c so it will be then easy to compare them (comparing manually resulted in : yes they match!)

openssl s_server does not seem to be able to produce the same trace (at least with my version of openssl(1.1.1g)) while the s_server code on their repository seems to be able to produce the same trace.

gilles-peskine-arm commented:
> However, what seems to be missing from those tests is the check that the SRTP keying material derivation on client and server lead to the same results.

Some ideas, not all practical:

* Have an option for the server to encrypt some fixed data with SRTP and have the client log it (or vice versa). If the keying material is not the same, the logged data from the client will be wrong.
* Have an option for the client to send extra data containing the keying material to the server as part of the payload, and have the server compare it, or vice versa.
* Run in fully deterministic mode and check known values in the logs. Problem: we don't currently have a fully deterministic mode in the test suite. The test programs support `reproducible=1`, which is fully deterministic as long as `MBEDTLS_PLATFORM_TIME_ALT` is set and `MBEDTLS_USE_PSA_CRYPTO` is unset (and perhaps other conditions that I'm forgetting), but we don't have that in `all.sh` and I'm not sure if `ssl-opt.sh` is fully geared for it.
* As you suggested, add an option to `run_test` to run a function that has access to both outputs.
* Do this testing as part of `test_suite_ssl` (or a new `test_suite_ssl_srtp`, maybe, since `test_suite_ssl` is getting really big). This doesn't allow interoperability testing, but it makes it easier to structure the testing without having to depend on shell interfaces.

hanno-arm commented:
> As you suggested, add an option to run_test to run a function that has access to both outputs.

I'd prefer this option since it's flexible and should hopefully be not too hard to implement. @jeannotlapin Would you like to give it a go? It would require to study the structure of `ssl-opt.sh` and, as @gilles-peskine-arm mentioned, likely add a variant of `run_test` which invokes a function on the logs `$CLI_OUT` and `$SRV_OUT`.

hanno-arm commented:
NB: The CI seems to be happy now :+1:

jeannotlapin commented:
> > As you suggested, add an option to run_test to run a function that has access to both outputs.
> 
> I'd prefer this option since it's flexible and should hopefully be not too hard to implement. @jeannotlapin Would you like to give it a go? It would require to study the structure of `ssl-opt.sh` and, as @gilles-peskine-arm mentioned, likely add a variant of `run_test` which invokes a function on the logs `$CLI_OUT` and `$SRV_OUT`.

I was thinking to add an option like -f ('-g'?) which would call the given script passing it the two logs as arg:
`

     "-g")
                if ! $2 "$SRV_OUT" "$CLI_OUT"; then

                    fail "function call to '$2' failed on Server and Client output"

                    return

                fi

                ;;
`

And add a dedicated script to parse both logs looking for a line starting by a given pattern (the first parameter of the script) and compare them.

hanno-arm commented:
@jeannotlapin This sounds good to me. @gilles-peskine-arm, are you fine with that, too?

mpg commented:
I just want to add that I think we should really have some interop testing for this feature if at all possible. Checking that our client and our server compute the same keying material is a good start and we should definitely do it, but it wouldn't detect situations where they would both compute the same wrong value.

jeannotlapin commented:
> I just want to add that I think we should really have some interop testing for this feature if at all possible. Checking that our client and our server compute the same keying material is a good start and we should definitely do it, but it wouldn't detect situations where they would both compute the same wrong value.

I'll modify the output of ssl_server2 to get a line with the same format produced by openssl s_client (see my previous post, the openssl s_server does not produce the output) so we can use the same script to at least check the that mbedtls server produces the same output as openssl client.

jeannotlapin commented:
@hanno-arm I added the tests, CI is happy... merging day?

jeannotlapin commented:
@hanno-arm, @mpg any more comments? Are we good for merging before it needs another rebase?

mpg commented:
@jeannotlapin Sorry for not replying last week... unfortunately there's still an important step before merging this: our policy is that a PR needs two approvals (in addition to the CI passing) before it can be merged. Since reviewing such as large PR (and again, thanks for making such a contribution!) is no trivial amount of work, it's something we'll need to plan - fortunately our next planning meeting is in two days, so hopefully we'll be able to allocate time for a second review in October.

As you say, we don't want to wait until there are conflicts to resolve again.

mpg commented:
(Btw Hanno is on leave, he's coming back next Wednesday, and I'm sure he'll review your last commit shortly after that.)

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -634,6 +634,23 @@ detect_dtls() {
+        fi
+
+        if grep "$srv_pattern" $3 >/dev/null; then :
+                       return 0;
```
Style: Indentation is off.

hanno-arm commented on programs/ssl/ssl_server2.c:
```diff
@@ -3884,7 +3884,17 @@ int main( int argc, char *argv[] )
+             * - compare client and server output
+             * - interop test with openssl which client produces this kind of output
+             */
+            mbedtls_printf( "    Keying material: " );
```
Not a blocker: I wonder if  it's worth making the new `-g` option a bit more flexible so that it's not necessary for server and client output to use _exactly_ the same output format. For example, `-g` might take two functions as arguments which are run on server/client output and whose output is then compared.  @mpg @gilles-peskine-arm ?

However, I think we should wait until the second review before changing anything - the current solution works, and the duplicated output isn't a major nuisance.

gilles-peskine-arm replied:
I feel YAGNI here. This is an internal interface and I don't expect a lot of pull requests in flight using the option, so we can change the semantics whenever we want.

hanno-arm reviewed (commented):
Thanks @jeannotlapin for extending the tests to compare the SRTP key materials in the OpenSSL compatibility test! Apart from a trivial style issue and the question of whether we'd want to make the new cli/srv output comparison option `-g` in `ssl-opt.sh` a bit more flexible, I'm happy with this PR and am ready to approve once the CI passes.

Could you rebase the PR on top of `development` to resolve the conflicts in `ssl_server2.c` and `ssl_client2.c`?

hanno-arm reviewed (approved):
Thanks for the rebase @jeannotlapin, it all looks good to me. The second review is still pending, but from my side once again thank you very much for this contribution and for your patience!

n-samir commented:
Thanks @jeannotlapin,  @hanno-arm and others for moving this forward.

lherman-cs commented:
@jeannotlapin @hanno-arm @gilles-peskine-arm and others, thanks a lot of driving this to completion!! Awesome work.

As the PR is already approved and it passes all the CI checks, are we waiting for something else for this PR to get merged?

yanesca commented:
@lherman-cs We require two approvals before merge, the second review is scheduled and will be completed soon.

yanesca reviewed (changes requested):
@jeannotlapin Thank you for your contribution and for your patience!

I think that the PR is in a very good shape, I only found a single compliance issue and two minor security issues. I also left comments for some minor typos, nitpicks, questions and suggestions.

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +869,35 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+
+#define MBEDTLS_TLS_SRTP_MAX_KEY_MATERIAL_LENGTH    60
```
It is linked to the size of the key and salt needed by the most demanding SRTP profile available (currently AES128, there is another RFC to use higher profiles but it will be another PR).

Actually this define has nothing to do in ssl.h since the key material generation now takes place in a callback (implemented in testers programs ssl_client2 and ssl_server2) so I'll remove this define to put it there.

yanesca replied:
Oh I see, thanks for explaining!

jeannotlapin commented on library/ssl_srv.c:
```diff
@@ -776,6 +776,127 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+    buf += 2;
+
+    /* check the buffer size: at least profiles + profile and mki length */
+    if( profile_length + size_of_lengths > len ||
```
profile_length is of type size_t but we use it to load a 16bits value read in the input buffer. So unless we built on a 16 bits processor having thus size_t mapped on a 16 bits unsigned integer, there is no possible overflow.

Is mbedtls being built on 16 bits processors? Unless this https://tls.mbed.org/discussions/crypto-and-ssl/help-with-ecc-and-bignum is no more relevant, I assume no.

I'll change it anyway just in case at some point 16 bits processors are supported.

gilles-peskine-arm replied:
It's ok to assume that `int` and `size_t` are 32-bit types or larger. There's already plenty of code in the library that assumes that adding a few 16-bit or 24-bit values doesn't overflow.

([The non-support for 16-bit platforms could stand to be better documented](https://github.com/ARMmbed/mbedtls/issues/3323).)

yanesca replied:
Yes, we can assume that `int` and `size_t` are 32 bit or larger. I missed that we load a 16 bit value in `profile_length`.

yanesca replied:
@jeannotlapin Thank you for changing it! This could save us a couple of false alarms in the future if someone makes the same mistake I did.

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -851,6 +869,35 @@ typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
+#define MBEDTLS_TLS_SRTP_MAX_KEY_MATERIAL_LENGTH    60
+#define MBEDTLS_TLS_SRTP_MAX_MKI_LENGTH             255
+#define MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH    4
+/*
```
mbedtls_ssl_srtp_profile used to be an enum, which I thought was safer and reassuring for the user but I've been told to change it...

I added the info to the description of mbedtls_ssl_get_dtls_srtp_protection_profile

jeannotlapin commented on library/ssl_cli.c:
```diff
@@ -1710,6 +1841,112 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     *                                      + length(2 bytes) + mki_len(1 byte)
+     *                                      and optional srtp_mki
+     */
+    if( ( len != 5 ) && ( len != ( 5 + mki_len ) ) )
```
as mki_len must be set to 0 if it is not used, the second condition is actually enough:
`( len != ( 5u + buf[4] ) )`

yanesca replied:
If we don't check `( len < 5 )` then accessing `buf[4]` might overread the buffer. I think it is probably better to make sure that `( len < 5 )` before accessing the byte at `buf[4]`.

jeannotlapin replied:
You're right, I just updated it

jeannotlapin commented on include/mbedtls/ssl.h:
```diff
@@ -3120,6 +3184,93 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ * \return         #MBEDTLS_TLS_SRTP_UNSET if the use of SRTP was not negotiated
+ *                 or peer's Hello packet was not parsed yet.
+ */
+mbedtls_ssl_srtp_profile mbedtls_ssl_get_dtls_srtp_protection_profile
```
Correct, there is no way to retrieve the mki value on server side.

I don't use the mki (and I don't think anyone does but it is part of the RFC so Ron decided to implement it), so it never occurred to me !

I'll change the API and improve the tests so we could actually check the mki value is the same on both sides. Next week.

jeannotlapin replied:
Ok it's done. The API now returns a pointer to the ssl context internal structure mbedtls_dtls_srtp_info.

I improved the tests so the mki value is extracted on both sides and compared. (this cannot be tested against others implementations as openssl does not [support yet the mki](https://github.com/openssl/openssl/issues/3435) and gnutls server does not seem to be able to produce traces of the srtp-dtls extension result, so we can just check that the server sends back the expected mki value)

yanesca commented on library/ssl_cli.c:
```diff
@@ -2278,6 +2515,16 @@ static int ssl_parse_server_hello( mbedtls_ssl_context *ssl )
+        case MBEDTLS_TLS_EXT_USE_SRTP:
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "found use_srtp extension" ) );
+
+            if( ( ret = ssl_parse_use_srtp_ext( ssl, ext + 4, ext_size ) ) != 0 )
```
@jeannotlapin You have marked this improvement as deployed, but I can't see it in your latest commit. Could you please check if you can see this there?

jeannotlapin replied:
I missed that one, thanks. It's actually done now

jeannotlapin commented:
Hi @yanesca,
 it shall be fine now. Let me know if the API returning the pointer to the selected profile and mki is ok for you.

yanesca commented on library/ssl_tls.c:
```diff
@@ -4751,10 +4751,10 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
+     mbedtls_ssl_get_dtls_srtp_negotiation_result( const mbedtls_ssl_context *ssl )
 {
-    return( ssl->dtls_srtp_info.chosen_dtls_srtp_profile );
+    return( &( ssl->dtls_srtp_info ) );
```
As far as I understand from the standard a typical use case is to do a handshake to establish an SRTP association, then immediately trigger a DTLS renegotiation several times to establish more SRTP associations for further data streams. In this case the `dtls_srtp_info` in the `ssl` context might be overwritten before the user accesses it. This would result in a silent error that depends on timing and is potentially very hard to debug, especially if there is concurrency involved.

I think it would be more robust to take a `mbedtls_dtls_srtp_info` output parameter as well and copy the negotiated values there.

Also, since we are copying already, we could check if the protection profile is `MBEDTLS_TLS_SRTP_UNSET` and if it is, set/leave `mki_len` to 0.

jeannotlapin replied:
No problem to change but if you're afraid the user won't call the get result function before a renegotiation and thus loose the first result, copying the output won't resolve it as the copy would be performed when the function is called so if a renegotiation is already ongoing we might have lost the result of the first handshake already.

The user could also copy the output content from the current version of the function, but if you feel more secure with copying it in the get function, it's fine for me.

yanesca replied:
If the user doesn't call the function before renegotiation then it can't really be helped. But if the user does call the get result function before a renegotiation, they still might access the returned pointer at some later point and by that time the pointed data will reflect the most recently negotiated protection profile and MKI.

You are right, the user can do the copying themselves, it is just more friendly and robust if we do it.

jeannotlapin replied:
I made the change to return a copied struct and not the pointer to the internal one

yanesca replied:
Thank you for the fast update and for your patience!

yanesca reviewed (changes requested):
I only have one suggestion to make the API slightly more robust.

yanesca commented on library/ssl_tls.c:
```diff
@@ -4751,10 +4751,16 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
 }
 
-const mbedtls_dtls_srtp_info *
+mbedtls_dtls_srtp_info
```
Most Mbed TLS functions don't return structures, usually they have output parameters for returning structures. This way the caller owns the memory management of the result and can decide if they wish to store it on the stack or on the heap. By output parameter I mean that the signature of the function could be like this:
```c
void mbedtls_ssl_get_dtls_srtp_negotiation_result( const mbedtls_ssl_context *ssl, mbedtls_dtls_srtp_info *dtls_srtp_info )
```

Could we please have an output parameter for the result here?

jeannotlapin replied:
no problem, here it is.

yanesca replied:
Thank you very much!

yanesca reviewed (approved):
Looks good to me.

yanesca commented:
@hanno-arm The PR is ready for your review again.

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -2509,8 +2511,11 @@ static int ssl_parse_server_hello( mbedtls_ssl_context *ssl )
 
-            if( ( ret = ssl_parse_alpn_ext( ssl, ext + 4, ext_size ) ) != 0 )
-                return( ret );
+            if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
```
This check normally happens in the extension handler, so that the top level structure of the message parser stays simple. Not a blocker.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -884,10 +885,7 @@ static int ssl_parse_use_srtp_ext( mbedtls_ssl_context *ssl,
-        {
-            ssl->dtls_srtp_info.mki_value[i] = buf[i];
-        }
+        memcpy(ssl->dtls_srtp_info.mki_value, buf, mki_length);
```
Nit: Spaces.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -2067,9 +2065,12 @@ static int ssl_parse_client_hello( mbedtls_ssl_context *ssl )
-                ret = ssl_parse_use_srtp_ext( ssl, ext + 4, ext_size );
-                if ( ret != 0 )
-                    return( ret );
+                if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
```
Same here: Ideally, the skip happens in the handler.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -2991,8 +2992,11 @@ static int ssl_write_server_hello( mbedtls_ssl_context *ssl )
-    ext_len += olen;
+    if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+    {
+        ssl_write_use_srtp_ext( ssl, p + 2 + ext_len, &olen );
```
Same here.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4727,16 +4727,14 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
          p++ )
     {
-        switch( *p )
+        if ( mbedtls_ssl_check_srtp_profile_value( *p ) != MBEDTLS_TLS_SRTP_UNSET )
```
Nit: Spacing.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -8874,12 +8875,14 @@ run_test  "DTLS-SRTP all profiles supported. server doesn't support mki." \
           -c "found srtp profile" \
           -c "selected srtp profile" \
           -c "DTLS-SRTP key material is"\
+          -c "DTLS-SRTP no mki value negociated"\
```
Typo: `s/negociated/negotiated/`

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4751,10 +4751,16 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
-    return( &( ssl->dtls_srtp_info ) );
+    mbedtls_dtls_srtp_info ret = ssl->dtls_srtp_info;
+    /* discard the mki if there is no chosen profile */
+    if ( ret.chosen_dtls_srtp_profile == MBEDTLS_TLS_SRTP_UNSET )
```
Nit: Space.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4751,10 +4751,16 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
      mbedtls_ssl_get_dtls_srtp_negotiation_result( const mbedtls_ssl_context *ssl )
 {
-    return( &( ssl->dtls_srtp_info ) );
+    mbedtls_dtls_srtp_info ret = ssl->dtls_srtp_info;
```
Nit: `ret` is always used for numeric error codes from what I remember. I'd prefer a different name. Certainly not a blocker.

jeannotlapin replied:
This is outdated code, last version does not use ret value as it gets an output argument

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -1277,6 +1396,19 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
 #endif
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
```
This logic normally happens in the extension writer. Not a blocker.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -1942,6 +2061,19 @@ static int ssl_parse_client_hello( mbedtls_ssl_context *ssl )
+            case MBEDTLS_TLS_EXT_USE_SRTP:
+                MBEDTLS_SSL_DEBUG_MSG( 3, ( "found use_srtp extension" ) );
+
+                if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
```
As above: This skipping logic would normally happen in the extension handler. Not a blocker.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -1942,6 +2061,19 @@ static int ssl_parse_client_hello( mbedtls_ssl_context *ssl )
+                if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+                {
+                    ret = ssl_parse_use_srtp_ext( ssl, ext + 4, ext_size );
+                    if ( ret != 0 )
```
Style: Spaces.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -2788,6 +2991,14 @@ static int ssl_write_server_hello( mbedtls_ssl_context *ssl )
 #endif
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
```
Style: Spaces.

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -2788,6 +2991,14 @@ static int ssl_write_server_hello( mbedtls_ssl_context *ssl )
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    if ( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+    {
```
As above: This skipping logic would normally be placed in the handler. Not a blocker.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4685,6 +4689,84 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    else
+    {
+        dtls_srtp_info->mki_len = ssl->dtls_srtp_info.mki_len;
+        memcpy( dtls_srtp_info->mki_value, ssl->dtls_srtp_info.mki_value, ssl->dtls_srtp_info.mki_len );
```
Style: Overly long line.

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -1710,6 +1842,121 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     */
+    for( i=0; i < ssl->conf->dtls_srtp_profile_list_len; i++)
+    {
+        if( server_protection == ssl->conf->dtls_srtp_profile_list[i] ) {
```
Style: `{` after line break.

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2597,7 +2749,74 @@ int main( int argc, char *argv[] )
+            }
+            else
+            {
+                mbedtls_printf( "    DTLS-SRTP no mki value negociated" );
```
`s/negociated/negotiated`

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -2240,6 +2336,35 @@ int main( int argc, char *argv[] )
+    {
+        if( opt.force_srtp_profile != 0 )
+        {
+            const mbedtls_ssl_srtp_profile forced_profile[] = { opt.force_srtp_profile, MBEDTLS_TLS_SRTP_UNSET };
```
Overly long line.

hanno-arm reviewed (changes requested):
Apart from trivial style issues, I'd prefer the skipping logic in case of DTLS to happen in the extension handlers - that's not a blocker, though. Otherwise: Ship it.

jeannotlapin commented:
@yanesca @hanno-arm I integrated the requested modifications, I hope it is fine now.

yanesca commented on programs/ssl/ssl_client2.c:
```diff
@@ -2597,7 +2750,74 @@ int main( int argc, char *argv[] )
+        mbedtls_ssl_get_dtls_srtp_negotiation_result( &ssl, &dtls_srtp_negotiation_result );
+
+        if( ( dtls_srtp_negotiation_result.chosen_dtls_srtp_profile
+                                == MBEDTLS_TLS_SRTP_UNSET ) )
```
The CI complains about the extra parentheses here.

jeannotlapin replied:
thanks, was not able to access the CI logs to see what it was complaining about. I just pushed the fix,

yanesca replied:
Yeah, we've been wanting to make the CI logs public for a long time, but it looks like that we still won't be able to do it in the near future :( 

Thank you for the fix!

yanesca commented on programs/ssl/ssl_server2.c:
```diff
@@ -3731,7 +3860,75 @@ int main( int argc, char *argv[] )
+        mbedtls_ssl_get_dtls_srtp_negotiation_result( &ssl, &dtls_srtp_negotiation_result );
+
+        if( ( dtls_srtp_negotiation_result.chosen_dtls_srtp_profile
+                                == MBEDTLS_TLS_SRTP_UNSET ) )
```
The CI complains about the extra parentheses here.

yanesca reviewed (changes requested):
The new changes look good to me, but the CI is failing and I left comments about the failures.

yanesca commented:
Thank you for your previous review @hanno-arm. The PR is ready for your review again.

yanesca reviewed (approved):
Looks good to me.

hanno-arm reviewed (approved):
Ship it! Thanks a lot @jeannotlapin for your work and patience.

yanesca commented:
Indeed, thank you @jeannotlapin for your work and patience.
